pkg/controller/replicaset/replica_set.go

```diff
@@ -209,10 +215,32 @@ func (rsc *ReplicaSetController) getPodReplicaSet(pod *api.Pod) *extensions.Repl
 		sort.Sort(overlappingReplicaSets(rss))
 	}
 
-	// update lookup cache
-	rsc.lookupCache.Update(pod, rss[0])
+	ret_rs = rss[0]
 
-	return rss[0]
+	// check ownerReference.Controller and UID to determine if the rc controls the pod
+	// in case of multiple controllers match the pod's label
+	if controllerRef != nil {
+		ret_rs = nil
+		for _, rs := range rss {
+			if controllerRef.UID == rs.ObjectMeta.UID {
+				ret_rs = rs
+				break
+			}
+		}
+	}
+
+	if ret_rs == rss[0] {
+		// update lookup cache only for the oldest controller
+		glog.V(4).Infof("Return for pod %s controller: %+v", pod.Name, *ret_rs)
+		rsc.lookupCache.Update(pod, ret_rs)
+	}
+
+	if ret_rs == nil {
+		// This should not happen
```

Comment:
We will hit this branch when user updates the Pod labels, specifically via this [call](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L345). We should return the oldest matching RS in this case, instead of nil. This RS will be enqueued and will later adopt the pod. (The old controlling RS is also [enqueued](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L340) and will reset the controllerRef of the pod)